download/tests/read-proxy-env.rs

```diff
@@ -0,0 +1,50 @@
+#![cfg(feature = "reqwest-backend")]
+
+use std::env::{remove_var, set_var};
+use std::sync::Mutex;
+
+use env_proxy::for_url;
+use lazy_static::lazy_static;
+use url::Url;
+
+lazy_static! {
+    static ref LOCK: Mutex<()> = Mutex::new(());
+}
+
+fn scrub_env() {
+    remove_var("http_proxy");
+    remove_var("https_proxy");
+    remove_var("HTTPS_PROXY");
+    remove_var("ftp_proxy");
+    remove_var("FTP_PROXY");
+    remove_var("all_proxy");
+    remove_var("ALL_PROXY");
+    remove_var("no_proxy");
+    remove_var("NO_PROXY");
+}
+
+// Tests for:
+// 1. Correctly retrieving the proxy (host, port) tuple from $https_proxy or $HTTPS_PROXY
+// 2. Prioritizing $https_proxy over $HTTPS_PROXY (except on Windows, where env vars
+//    are case-insensitive)
+#[test]
+fn read_basic_proxy_params() {
+    let _l = LOCK.lock();
+    scrub_env();
+    set_var("https_proxy", "http://proxy.example.com:8080");
+    #[cfg(not(windows))]
+    set_var("HTTPS_PROXY", "http://proxy2.example.com:8080");
+    let u = Url::parse("https://www.example.org").ok().unwrap();
+    assert_eq!(
+        for_url(&u).host_port(),
```

Comment:
This is using the env_proxy function directly: I guess thats what broke, so we probably need to, but its a bit of a shame to be testing our dependencies :(. I haven't looked but if we wrap things up in higher order things maybe those should be checked instead/too.

Reply:
Looking at the antecedents of #2399, it should be clear that `env_proxy`'s own logging and tests _did_ detect the breakage, but it wouldn't have helped here; the even higher order problem stemmed from my failure to publish the updated version in a timely manner.

I agree that a higher-level test would be architecturally cleaner, and could imagine spinning up a separate micro-proxy thread (could it be backed by `hyper`? Probably) and attempting to fetch a file through that. In any case, proxies are an important enough use case for `rustup` that I believe that they should be tested for in some manner.